mesh/meshdb.go

```diff
@@ -327,30 +350,168 @@ func (m *MeshDB) writeTransactions(txs []*types.AddressableSignedTransaction) er
 	return nil
 }
 
-func (m *MeshDB) GetTransactions(transactions []types.TransactionId) (
-	map[types.TransactionId]*types.AddressableSignedTransaction,
-	[]types.TransactionId) {
+func (m *MeshDB) addToMeshTxs(txs []*types.Transaction, layer types.LayerID) error {
+	groupedTxs := groupByOrigin(txs)
 
-	var mIds []types.TransactionId
-	ts := make(map[types.TransactionId]*types.AddressableSignedTransaction, len(transactions))
-	for _, id := range transactions {
-		t, err := m.GetTransaction(id)
-		if err != nil {
-			m.Warning("could not fetch tx, %v %v", hex.EncodeToString(id[:]), err)
-			mIds = append(mIds, id)
-		} else {
-			ts[id] = t
+	for addr, accountTxs := range groupedTxs {
+		if err := m.addToAccountTxs(addr, accountTxs, layer); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (m *MeshDB) addToAccountTxs(addr types.Address, accountTxs []*types.Transaction, layer types.LayerID) error {
+	m.meshTxsMutex.Lock()
+	defer m.meshTxsMutex.Unlock()
```

Comment:
why do we need to lock everything ?
cant we lock on the specific account ?

Reply:
1. Only _this part_ of `addBlockWithTxs` will be serial.
2. This is temporary, after merging this PR I'm going to change the implementation to not require a lock at all (by using LDB iteration to manage the list).
3. While not rocket science, implementing this isn't totally trivial.
4. I wouldn't want to have a temporary bug, but a temporary potential (unproven) performance issue, that's not trivial to fix—I think we can live with.